woocommerce/woocommerce-ios
# Edit Grouped Products: edit linked products (Part 1/3) - paginated list of linked products with delete action

Part of #2199

- [x] ⚠️ Update PR base to `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2492 is merged ⚠️ 

This PR implements the highlighted screen in the flow to edit linked products for a grouped product:

![grouped_products_pr](https://user-images.githubusercontent.com/1945542/87378495-7a08a100-c5c0-11ea-8d0c-344594a0f55e.png)

## Changes

- Networking layer:
  -  I later realized the API for fetching products given a list of IDs still requires pagination, thus added `pageNumber: Int` and `pageSize: Int` params to `ProductsRemote.loadProducts(siteID:productIDs:...)`
  - Included `groupedProducts` when encoding a `Product` for remote update
- Yosemite layer:
  - Similar to the Networking layer, added `pageNumber` and `pageSize` to `ProductAction.retrieveProducts(siteID:productIDs:...)` and in `ProductStore`
- App layer:
  - Added a new icon `deleteCellImage: UIImage` using Gridicons' `cross` icon for deleting a linked product + test case
  - Created `GroupedProductsViewController` with a xib that contains a CTA to add more products (no-op in this PR), and a container view of a paginated list of linked products via a child view controller `PaginatedListSelectorViewController` & `GroupedProductListSelectorDataSource`
  - `GroupedProductListSelectorDataSource` configures the results controller to show only the linked products in the order when they were added, configures each product cell, syncs the next page of products, and handles actions like deleting a linked product and adding more products (in a future PR). It also keeps track of the latest linked product IDs with an observable
  - In `ProductFormViewController`, navigated to `GroupedProductsViewController` when the user taps on the "Grouped Products" row and saves the grouped products on completion
  - 🔆 (trickier part) Changes to `PaginatedListSelectorViewController`/`PaginatedListSelectorDataSource`. Since it is currently used for a fixed paginated list, some changes are required to support this PR use case where the user can delete any number of products from this list:
    - Added `func updateResultsController()` to refresh the results controller when it changes in the data source. In this PR, this is called by `GroupedProductsViewController` when the list of observable product IDs changes
    - Added an optional getter `customResultsSortOrder` to `PaginatedListSelectorDataSource` protocol for sorting the results in a way that Core Data doesn't currently support. In this PR use case, the linked products are displayed in the same order as `product.groupedProducts` array. First I tried passing a `comparator` to `NSSortDescriptor` for `ResultsController`, however `comparator` is not supported in Core Data (it'd crash the app from using a `comparator` in SQLite fetch) ([SO post](https://stackoverflow.com/questions/14938653/unsupported-nssortdescriptor-comparator-blocks-are-not-supported) on this). Therefore, I added an optional sort strategy to `PaginatedListSelectorDataSource`, and `PaginatedListSelectorViewController` sorts the fetched objects when it is rendering a cell given the index path. I understand that this isn't a performant solution, and am open to any ideas to sort the fetched objects from a `ResultsController`!

## Testing

Prerequisite: the site has a grouped product with non-zero linked products

- Go to the Products tab
- Tap on the grouped product in the prerequisite
- Tap on the "Grouped products" row --> a list of linked products should be displayed with delete CTA for each product
- Tap on "X" to delete a product --> "Done" CTA should appear in the navigation bar
- Tap "Done" to go back to product details
- Tap "Update" to update the product --> the grouped product's linked products should be updated remotely

## Example screenshots

state | light | dark
-- | -- | --
results | ![Simulator Screen Shot - iPhone 11 - 2020-07-14 at 13 05 33](https://user-images.githubusercontent.com/1945542/87386496-2784b000-c5d3-11ea-8b47-95cc884b2761.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-14 at 13 06 02](https://user-images.githubusercontent.com/1945542/87386500-2a7fa080-c5d3-11ea-9c1d-46bbb3f7aca0.png)
empty | ![Simulator Screen Shot - iPhone 11 - 2020-07-14 at 13 08 31](https://user-images.githubusercontent.com/1945542/87386502-2bb0cd80-c5d3-11ea-9d63-8c96f973f56f.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-14 at 13 08 23](https://user-images.githubusercontent.com/1945542/87386501-2bb0cd80-c5d3-11ea-9a45-63f794670eb0.png)


Update release notes:
